### ISSUE TITLE
Optimalizace aktivní vrstvy

### ISSUE THREAD

**Issue by severak** [osmcz/osmcz#100]
Koukal jsem se na mapy.cz, čím to, že je jejich aktivní vrstva tak rychlá.

Došel jsem k následujícímu závěru: Data pro vrstvu netahají po dlaždicích, ale najednou pro celý BBOX (request `poiagg`). Toto se volá pokaždé, když uživatel mapu posune, změní zoom nebo se změní viewbox. Requesty jsou pravděpodobně kódovány pomocí seznamáckého [fastRPC](http://jak.seznam.cz/doc/symbols/JAK.FRPC.html#constructor).

Hlavní trik je myslím načítat to za celý BBOX. Druhý trik je umět rychle nahodit nová data a uklidit stará. Třetí trik už jsme řešili - různá POI pro různé zoomy.

**Comment by mkyral** [comment]
Takže hlavní je získat data v kompaktní formě a co nejrychleji je nahrát?

S js moc zkušenosti nemám, ale podle mne nejdéle trvá právě ten úklid. Nejlépe to je vidět, při zoomování - nové dlaždice se načítají postupně, ale to zahození staré vrstvy trvá docela dost dlouho.

**Comment by severak** [comment]
Divný. Zkoušel jsem to projíždět profilerem ve firebugu a ten mi tvrdil, že nejvíce zdržuje [tato funkce](https://github.com/osmcz/osmcz/blob/b7656344e3240ecac9ac2227c926c17f65c51db0/js/osmcz.js#L54).

**Comment by Krakonos** [comment]
Kde tato funkce je initmap(), nikoliv volání toho map(). Spíš mi přijde zvláštní, kolikrát se volá. Ale jsou tam nějaké horší věci, jako GeoJSON.addTileData. Tam se manipuluje s každým objektem zvlášť (nebo to tak vypadá) a není zrovna triviální.

Myslím, že tohle řešení bude to, co musíme opravit. Líbí se mi myšlenka toho, jak to má geocaching (někdo tady správně zmínil), že jsou předrenderované dlaždice overlaye a tudíž javascript nemusí vůbec kreslit ikonky na mapu (což je imho nejvíc práce na celé vrstvě).

**Comment by mkyral** [comment]
Akorát, že s mapou geocachingu mám občas problém, že ikonka sice je, ale nedá se na ní kliknout, protože data ještě nedotekla a chybí jakákoli signalizace, že se tak již stalo. Takže mám ikonku, ale ne a ne ji rozkliknout - někdy na zabití.

**Comment by severak** [comment]
Mapy.cz mají klikatelné POI mimo samotné mapové dlaždice. 

A jak koukám - mají jich podstatně méně než OSM aktivní vrstva. Např. na Karlově mostě máme POI pro každou sochu, kdežto Seznam má jen jedno (unesco) POI za celý Karlův most. _To je asi ten hlavní trik._

**Comment by Krakonos** [comment]
@severak Jo. Protože to mají vyladěné. Takové věci můžeme také nějak zařídit, ale určitě nemůžeme spoléhat na to, že filtrujeme dobře a když tomu tak někde na světě nebude, celé to poletí do kytek...

@mkyral Souhlasím, také mě to irituje. To se dá ošetřit třeba tak, že zobrazíš dlaždici až když k ní máš klikací data.

**Comment by zbycz** [comment]
Myslím si že hlavní rozdíl proti seznamu je množství dat. To že načítají
BBOX viewportu je dle mého ve výsledku pomalejší než dlaždice.  (srovnej s
obrázkem jako bbox / dlaždice).

Vyladit to samozřejmě jde - někdo třeba zkuste.
- vyzkoušet různé ikonky - SVG/PNG/GIF
- poladit funkci tiled-geojsonu, aby efiktivněji manipuloval s daty

**Comment by Krakonos** [comment]
@zbycz Trochu. Co jsem se díval, tak u nás to vypadá tak, že se pošle třeba 30 requestů, každý z nich vrátí nějaký triviální geojson co má tak 50B. To na rychlém připojení je úplně jedno, myslím si, že třeba na mobilu bude výrazně horší, kvůli odezvám.

**Comment by zbycz** [comment]
A kde vlastně spatřujeme "pomalost" - i obecně všude, nebo zejména v tom
centru Prahy? Abych věděl, co je vlastně za problém :-)

**Comment by Krakonos** [comment]
@zbycz Prakticky kdekoliv, kde je více POI na jednom místě. Včera jsem se díval na Svobodu nad úpou a Horní Maršov.

Pravda je, že se na to dívám na velkém monitoru, ale nic extra v dnešní době (1920x1080-trošku okrajů nahoře a dole).

**Comment by mkyral** [comment]
Tak mně funguje nejhůř v centru Prahy. Ale mám docela dost výkonnný NB se 16GB RAM. Obecně je nejhorší zoomování - to se mi to zasekne na několik (i desítek) sekund. Posun mapy do stran je celkem v pohodě.

**Comment by Krakonos** [comment]
Centrum prahy je zabijácké. Čekání řádově 10s je ale celkem běžné i v mnohem řidčích místech...

**Comment by zbycz** [comment]
Zkusil jsem udělat profiling javascriptu a kupodivu to čeká na dost divných místech:
![screen shot 2016-04-13 at 12 01 58](https://cloud.githubusercontent.com/assets/385047/14489855/c72ebdf8-016f-11e6-8641-35b3682375ba.png)

Jako částečné řešení navrhuju do `TileLayer.GeoJSON.js` přidat pro zoomy 15-17 přidat nějaký filtr, že pokud je v dlaždici příliš POIek, tak to nějaké zahodí.
1. buď náhodně
2. nebo třeba vyhodí amenity, těch bývá dost
3. nebo navrhněte? :-)

Myslím, že problém latence bych vůbec neřešil - load je u mě subjektivně trochu pomalý nevadí. Co vadí hodně, že mapa nejde zoomovat a posouvat. Souhlas?
